### ISSUE TITLE
Abnormal YOLOv8 validation mAP jump after resuming training

### ISSUE THREAD

**Issue by WenTheProgrammer** [ultralytics/ultralytics#6712]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I wonder why sometimes after I stopped training a model and then continued training the model, the validation mAP50s would jump to a much higher value. I have seen this for both YOLOv8n and YOLOv8x so maybe it's not model specific. Attached are the results.csv file saved during training and a plot showing mAP50(B), where you can clearly see the abnormal jumps when I stopped and resumed training. Any help is appreciated!
[results.csv](https://github.com/ultralytics/ultralytics/files/13518693/results.csv)
![image](https://github.com/ultralytics/ultralytics/assets/67032259/6f9b52a2-1bea-4846-bedb-e3fe041bf5ab)

**Comment by glenn-jocher** [comment]
@WenTheProgrammer hello! It's great to see your active involvement with the YOLOv8 training process. The phenomenon you're experiencing with the mAP50 jumps after resuming training could be related to several factors:

1. **Learning Rate Scheduling**: If your training configuration involves a learning rate scheduler, pausing and resuming training might interact with the schedule differently than a continuous training run. This could cause the learning rate to be in an unexpected state upon resumption, temporarily affecting performance.

2. **Model State Restoration**: Ensure the model's state, including the weights, optimizer state, and learning rate scheduler, is being accurately restored when resuming training. Inaccuracies in state restoration could lead to such jumps in performance metrics.

3. **Data Loader State**: Similarly, the state of the data loader (like data augmentations and shuffle patterns) can influence the validation results. If the validation loader starts from a different point or with different settings, this might temporarily affect the mAP.

4. **Batch Normalization**: The batch normalization statistics can be affected when training is paused and resumed, particularly if there are differences in batch sizes or the number of batches used for tracking running statistics.

5. **Random Seed Set**: Check if you're setting a random seed that could impact the initialization of certain aspects of training after resumption.

While these are some general suggestions, each training scenario can have unique aspects. If necessary, consult the Ultralytics Docs for more detailed information on training configuration and troubleshooting to ensure a smoother training and validation process. If the issue persists, please provide more context so the community can further assist you. Happy training! 😊🚀

**Comment by WenTheProgrammer** [comment]
Hi @glenn-jocher, thank you for the answer! The code below is how I started and resumed a training session:
```
from ultralytics import YOLO

# Start training
model = YOLO('yolov8n.pt')
results = model.train(data="data.yaml", epochs=300, batch=16, imgsz=480)

# Resume training
model = YOLO('last.pt')
# I noticed specifying epochs and imgsz when resuming training is unnecessary
results = model.train(data="data.yaml", batch=16, resume=True)
```
I did set a constant random seed when selecting my training, validation, and test datasets so they never change. 
Do you observe any error in my code that might cause the mAP50 jump issue? Much appreciate your help!

**Comment by glenn-jocher** [comment]
@WenTheProgrammer, it looks like you're following the correct steps for starting and resuming training using the YOLO class. The code to resume training with `'last.pt'` and setting `resume=True` is generally sound. Here are a few subtleties to consider:

1. **Ensure Consistent Environment**: Confirm that the training and validation environment is consistent between sessions. This includes the version of the codebase, dependencies, and hardware utilized if possible.

2. **Review Training Logs**: Make sure to review the training logs for any messages or warnings related to the training state that could provide clues about the jumps in performance.

3. **Random Seed**: Even though you have set a random seed for dataset selection, check if you need to set it again for the training itself to ensure reproducibility.

4. **Optimizer State**: Keep in mind that apart from the model weights in `last.pt`, the optimizer state is also crucial for continuous training performance. If the optimizer state is not saved or loaded correctly, this could lead to fluctuations in performance.

5. **Batch Normalization Statistics**: As mentioned earlier, watch out for any potential impacts on batch normalization layers when resuming training, as these can impact the model's validation performance.

Given the correct code implementation, the issue could likely be related to the intricacies of the training procedure rather than the code itself. If none of these suggestions resolve your concern, you might want to perform additional steps like a thorough review of the configuration files and the training outputs. If there are still anomalies, consider reaching out on the Ultralytics Discussions to see if other users can share insights from similar experiences. Keep up the meticulous work! 🙂🔍

**Comment by WenTheProgrammer** [comment]
Hi @glenn-jocher, just an update on this issue. I thought the issue came from not setting a random seed when I create my training, validation, and test datasets, even though I am pretty sure I did, so I trained a new model with a constant random seed, imgsz of 640, epoch of 300, and batch of 64 to see how things go. I used the same script to train the model in Google Colab. I resumed training the model multiple times, everything went well and I never observed the abnormal mAP jump, until today. Last night I stopped training as my validation mAP has plateaued, which was around 0.69. Today I checked what the mAP of my best weights on my test dataset was using `model.val()`, and it was around 0.79. I couldn't understand why my test accuracy is so much higher than my validation accuracy, so I checked my validation dataset again using my best weights and `model.val()`, and it also turned out to be around 0.79, but it was round 0.69 yesterday!? I tried to resume training again, using my latest weights, and of course, the mAP jump issue happened again, jumping from 0.69 to 0.79.

So just to summarize everything, the same model weights, the same validation dataset, the same configuration file, the same script, yesterday the model mAP was 0.69, today the model mAP was 0.79. The issue apparently happens only occasionally instead of all the time. I cannot think of anything that would possibly cause this issue and I am very much puzzled...

I attached "results.csv" in case it's helpful.
[results.csv](https://github.com/ultralytics/ultralytics/files/13554098/results.csv)

"args.yaml" cannot be attached so I'm providing it here:

task: detect
mode: train
model: runs/detect/train2/weights/last.pt
data: /content/train_data.yaml
epochs: 300
patience: 50
batch: 64
imgsz: 640
save: true
save_period: -1
cache: false
device: null
workers: 8
project: null
name: train2
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: false
opset: null
workspace: 4
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.0
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
cfg: null
tracker: botsort.yaml
save_dir: runs/detect/train2

**Comment by glenn-jocher** [comment]
@WenTheProgrammer, it's certainly puzzling when mAP values show such unexpected changes. Since you've maintained consistency with your model configurations and random seed settings, let's consider other factors that might affect the observed mAP discrepancy:

1. **Google Colab Environment**: Note that Google Colab environments can reset upon session termination, potentially affecting reproducibility if the entire environment, including library versions and installation states, isn't recreated identically each time.

2. **Caching Mechanism**: Ensure `cache=false` isn't leading to any variability in data loading between sessions. Although caching is disabled, small environmental differences could still influence data loading upon each restart.

3. **Model Weight Stability**: Check the stability and behavior of the batch normalization layers within your model. Their internal states can change between training and evaluation, especially after resuming training.

4. **Hardware Acceleration**: Keep an eye on the use of Mixed Precision Training (`amp=true`) as it can sometimes introduce non-determinism, albeit usually minor.

5. **Validation/Test Set Handling**: Double-check the handling of your validation and test sets to confirm there's no inadvertent mixing or mislabeling that could misrepresent performance metrics.

For a more in-depth diagnosis, you might want to reach out on Ultralytics Discussions, providing your "results.csv" and "args.yaml" for additional community support. It's cases like these that can be quite challenging, but they also push us towards more robust solutions. Hang in there! 🛠️🔍

**Comment by WenTheProgrammer** [comment]
Hi @glenn-jocher, issue finally solved! The problem came from `random.seed()` function. I used it to create the "same" training, validation, and test datasets every time when I execute the code in a new runtime in Colab. It turned out the function would create different datasets from time to time. The lesson learned here is simply saving the dataset image lists to files when the code is executed for the first time, and loading the files for future code executions...

**Comment by glenn-jocher** [comment]
@WenTheProgrammer, fantastic news! I'm glad to hear that you pinpointed the issue to the `random.seed()` usage and its inconsistencies. Saving dataset splits to files is indeed a reliable way to ensure you're working with the same data across different training sessions. Thank you for sharing your solution—it's a valuable lesson for reproducibility in ML experiments. Good luck with your further work! 👍🎉
